### DISCUSSION TITLE
Implementing parsing of coordination compounds drawn following IUPAC rules.

### DISCUSSION THREAD

**Discussion by hrp1000** [rdkit/rdkit#6019] [Development]
Hi

I have a problem with this entry. If I try to generate a fingerprint for calculating the Tanimoto coefficient with CHEBI:60344 with this code, it works fine for CHEBI:17267, but fails for CHEBI:26355 -

>     from rdkit import Chem, DataStructs
>     from bioservices import ChEBI
>     heme = ChEBI()
>     heme_chebi_id = "CHEBI:60344"
>     heme_smiles = heme.getCompleteEntity(heme_chebi_id).smiles
>     target = Chem.MolFromSmiles(heme_smiles)
>     fp2 = Chem.RDKFingerprint(target)
>     for chebi_id in ["CHEBI:17627", "CHEBI:26355"]:
>       ch = ChEBI()
>       smiley = ch.getCompleteEntity(chebi_id).smiles
>       print("reference:", heme_chebi_id)
>       print("target: ", chebi_id)
>       print("reference:", heme_smiles)
>       print("target: ", smiley)
>       ref = Chem.MolFromSmiles(smiley)
>       fp1 = Chem.RDKFingerprint(ref)
>       Tan = DataStructs.TanimotoSimilarity(fp1, fp2)
>       print(Tan)
>       print("-" * 64)
>     exit()
> 

the output from this clearly shows a mis-match between what is supplied from the EBI and what RDKit expects - 

>     reference: CHEBI:60344 target: CHEBI:17627 reference: CC1=C(CCC([O-])=O)C2=[N+]3C1=Cc1c(C)c(C=C)c4C=C5C(C)=C(C=C)C6=[N+]5[Fe--]3(n14)n1c(=C6)c(C)c(CCC([O-])=O)c1=C2
> 
>     target: CC1=C(CCC(O)=O)C2=[N+]3C1=Cc1c(C)c(C=C)c4C=C5C(C)=C(C=C)C6=[N+]5[Fe--]3(n14)n1c(=C6)c(C)c(CCC(O)=O)c1=C2 Tanimoto coefficient: 1.0
> 
>     reference: CHEBI:60344 target: CHEBI:26355 reference: CC1=C(CCC([O-])=O)C2=[N+]3C1=Cc1c(C)c(C=C)c4C=C5C(C)=C(C=C)C6=[N+]5[Fe--]3(n14)n1c(=C6)c(C)c(CCC([O-])=O)c1=C2 target: CC1=C(CCC(O)=O)C2=[N]3C1=Cc1c(C)c(C=C)c4C=C5C(C)=C(C=C)C6=[N]5[Fe]3(n14)n1c(=C6)c(C)c(CCC(O)=O)c1=C2 [12:36:26] Explicit valence for atom # 9 N, 4, is greater than permitted Traceback (most recent call last): File "/Volumes/Users/harry/icl/phyre2-ligand/./tanimoto_test.py", line 20, in fp1 = Chem.RDKFingerprint(ref) Boost.Python.ArgumentError: Python argument types in rdkit.Chem.rdmolops.RDKFingerprint(NoneType) did not match C++ signature: RDKFingerprint(RDKit::ROMol mol, unsigned int minPath=1, unsigned int maxPath=7, unsigned int fpSize=2048, unsigned int nBitsPerHash=2, bool useHs=True, double tgtDensity=0.0, unsigned int minSize=128, bool branchedPaths=True, bool useBondOrder=True, boost::python::api::object atomInvariants=0, boost::python::api::object fromAtoms=0, boost::python::api::object atomBits=None, boost::python::api::object bitInfo=None)
> 

I have contacted the EBI/Chebi about this and got the reply:

> Historically, coordination bonds have been depicted in a variety of ways but IUPAC recommends that these bonds should be depicted as regular 'plain' bonds as shown in CHEBI:26355 (see other examples: https://iupac.qmul.ac.uk/tetrapyrrole/TP8.html)
> 
> Some existing software including RDKit are unable interpret properly coordination bonds with single bonds and without charges since the structure does not satisfy their strict valence criteria so therefore is unable to generate a fingerprint. You can either contact RDKit about this issue or add charges to the structure to satisfy RDKit's criteria.
> 

So clearly they think it's an issue with RDKit being too strict and not adhering to what IUPAC recommends.

What do you think?

**Comment by greglandrum** [comment]
I actually wasn't aware of the IUPAC recommendations on this one, thanks for providing that pointer.
The recommendation is unfortunate (because I think it's ugly), but it's a good argument for us to add a special case to the `cleanup()` function to recognize atoms with single bonds to metal atoms with a valence that's one too high and clean that up.

My suggestion would be that we add something to recognize atoms which have a calculated valence which is X too high but which have X single bonds to metal atoms and to replace those single bonds with coordination bonds.
So this:
![image](https://user-images.githubusercontent.com/540511/214849304-6a82a7cc-607a-4227-bcab-3adf397a2520.png)
would be converted to this:
![image](https://user-images.githubusercontent.com/540511/214849366-d99de050-59c4-4242-ac8e-d16e97f40f8d.png)

Note that I think this will not work for heteroatoms which are aromatic in the input. For example in this structure:
![image](https://user-images.githubusercontent.com/540511/214849804-f8dede1f-391c-4e57-b530-d37aa44789eb.png)

There would be no way to know which two N atoms should have coordination bonds, so those would continue to generate errors. This wouldn't be a problem with RDKit generated SMILES since there the bonds would be coordinate to start with.

@DavidACosgrove @d-b-w : I know that you both have put some thought into coordination/organometallic systems. What do you think?

**Comment by hrp1000** [comment]
One thing that would really help people like me, would be for a Python exception to be raised when things like this are encountered (there are bound to be other problems that have not yet been raised with RDKit developers); I haven't found a way to catch the Boost.Python.ArgumentError (C++?) exception inside my Python code, but that may just be my inexperience!

**Reply by greglandrum** [reply]
You don't need to catch an exception. Just check for None being returned by `MolFromSmiles()`

**Reply by hrp1000** [reply]
THX!

**Comment by DavidACosgrove** [comment]
I agree that the IUPAC's recommendation is unhelpful.  That sort of bond seems different from a single bond to me, though the whole organometallics area doesn't work especially well in the valence bond model, IMO, and I guess this is just another example.  I think @greglandrum's suggestion makes sense and is probably the easiest way of having RDKit cope with this sort of molecule.

**Comment by d-b-w** [comment]
I agree with your suggestion, Greg. I also checked with someone on sdgr's materials team, who pointed out that this is exactly what we do internally.. :)

**Reply by greglandrum** [reply]
Glad to hear that!
